File: appkit-creating-and-customizing-the-touch-bar/Objective-C/NSTouchBar Catalog/BackgroundWindow/PhotoManager.m
/*
See the LICENSE.txt file for this sample’s licensing information.

Abstract:
Singlelton object for asynchronously loading all photos from the Desktop Pictures folder.
*/

@import Cocoa;
@import Foundation;

#import "PhotoManager.h"

NSString *kImageNameKey = @"name";
NSString *kImageKey = @"image";

@implementation PhotoManager: NSObject

+ (PhotoManager *)shared
{
    static PhotoManager *photoManager;  // The singleton PhotoManager controller.
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        photoManager = [[PhotoManager alloc] init];
    });
    return photoManager;
}

- (instancetype)init
{
    self = [super init];
    if (self != nil)
    {
        dispatch_async(dispatch_get_global_queue(QOS_CLASS_BACKGROUND, 0), ^{
            self.photos = [NSMutableArray array];
            for (NSUInteger index = 1; index <= 14; index++)
            {
                NSString *imageName = [NSString stringWithFormat:@"image%ld", index];
                NSImage *fullImage = [NSImage imageNamed: imageName];
                NSSize imageSize = [fullImage size];
                if (imageSize.width > 0 && imageSize.height > 0)
                {
                    CGFloat thumbnailHeight = 30;
                    NSSize thumbnailSize = NSMakeSize(ceil(thumbnailHeight * imageSize.width / imageSize.height), thumbnailHeight);
                    
                    dispatch_async(dispatch_get_main_queue(), ^{
                        NSImage *thumbnail = [[NSImage alloc] initWithSize:thumbnailSize];
                        
                        [thumbnail lockFocus];
                        [fullImage drawInRect:NSMakeRect(0, 0, thumbnailSize.width, thumbnailSize.height) fromRect:NSMakeRect(0, 0, imageSize.width, imageSize.height) operation:NSCompositingOperationSourceOver fraction:1.0];
                        [thumbnail unlockFocus];

                        NSMutableDictionary *dict = [NSMutableDictionary dictionaryWithObjectsAndKeys:
                                                     [imageName lastPathComponent], kImageNameKey,
                                                     thumbnail, kImageKey,
                                                     // You can add any additional pertinent information for this image object here.
                                                     nil];
                        [self.photos addObject:dict];
                    });
                }
                
            }

            self.loadComplete = YES;
            
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.delegate didLoadPhotos:self.photos];
            });
        });
    }
	return self;
}

@end

File: appkit-creating-and-customizing-the-touch-bar/Swift/NSTouchBar Catalog/PhotoManager.swift
/*
See the LICENSE.txt file for this sample’s licensing information.

Abstract:
Singlelton object for asynchronously loading all photos from the image assets.
*/

import Cocoa
import Foundation

final class PhotoManager {
    
    static let ImageNameKey = "name"
    static let ImageKey = "image"
    
    static let shared = PhotoManager()
    
    weak var delegate: PhotoManagerDelegate?
    
    var photos = [Any]()
    var loadComplete = false
    
    private init() {
        DispatchQueue.global(qos: .background).async {
            for index in 1...14 {
                let imageName = "image" + "\(index)"
                if let fullImage = NSImage(named: imageName) {
                    let imageSize = fullImage.size
                    
                    guard imageSize.width > 0 && imageSize.height > 0 else { continue }
                    
                    let thumbnailHeight: CGFloat = 30
                    let thumbnailSize = NSSize(width: ceil(thumbnailHeight * imageSize.width / imageSize.height), height: thumbnailHeight)
                    
                    DispatchQueue.main.async {
                        let thumbnail = NSImage(size: thumbnailSize)
                        thumbnail.lockFocus()
                        fullImage.draw(in: NSRect(origin: .zero, size: thumbnailSize),
                                       from: NSRect(origin: .zero, size: imageSize),
                                       operation: .sourceOver,
                                       fraction: 1.0)
                        thumbnail.unlockFocus()

                        var imageDict = [String: Any]()
                        imageDict = [PhotoManager.ImageKey: thumbnail as Any,
                                     PhotoManager.ImageNameKey: imageName]
                        self.photos.append(imageDict)
                    }
                }
            }

            self.loadComplete = true
            DispatchQueue.main.async {
                self.delegate?.didLoadPhotos(photos: self.photos)
            }
        }
    }
}

protocol PhotoManagerDelegate: AnyObject {
    func didLoadPhotos(photos: [Any])
}
